torch/lib/c10d/TCPStore.cpp

```diff
@@ -0,0 +1,352 @@
+}
+
+void TCPStoreDaemon::run() {
+  // Create the control pipe
+  if (pipe(controlPipeFd_.data()) == -1) {
+    throw std::runtime_error("Failed to create the control pipe to start the "
+                             "TCPStoreDaemon run");
+  }
+
+  std::vector<struct pollfd> fds;
+  fds.push_back({ .fd = storeListenSocket_, .events = POLLIN });
+  // Push the read end of the pipe to signal the stopping of the daemon run
+  fds.push_back({ .fd = controlPipeFd_[0], .events = POLLHUP });
+
+  // receive the queries
+  bool finished = false;
+  while (!finished) {
+    for (size_t i = 0; i < sockets_.size(); i++) {
+      fds[i].revents = 0;
+    }
+
+    SYSCHECK(::poll(fds.data(), fds.size(), -1));
+
+    // TCPStore's listening socket has an event and it should now be able to
+    // accept new connections.
+    if (fds[0].revents != 0) {
+      if (fds[0].revents ^ POLLIN) {
+        throw std::system_error(ECONNABORTED, std::system_category(),
+            "Unexpected poll revent on the master's listening socket: " +
+            std::to_string(fds[0].revents));
+      }
+      int sockFd = std::get<0>(tcputil::accept(storeListenSocket_));
+      sockets_.push_back(sockFd);
+      fds.push_back({ .fd = sockFd, .events = POLLIN });
+    }
+    // The pipe receives an event which tells us to shutdown the daemon
+    if (fds[1].revents != 0) {
+      // Will be POLLUP when the pipe is closed
+      if (fds[1].revents ^ POLLHUP) {
+        throw std::system_error(ECONNABORTED, std::system_category(),
+            "Unexpected poll revent on the control pipe's reading fd: " +
+            std::to_string(fds[1].revents));
+      }
+      finished = true;
+      break;
+    }
+    // Skipping the fds[0] and fds[1],
+    // fds[0] is master's listening socket
+    // fds[1] is control pipe's reading fd
+    for (size_t fdIdx = 2; fdIdx < fds.size(); ++fdIdx) {
+      if (fds[fdIdx].revents == 0) {
+        continue;
+      }
+
+      if (fds[fdIdx].revents ^ POLLIN) {
+        throw std::system_error(ECONNABORTED, std::system_category(),
+            "Unexpected poll revent: " +
+            std::to_string(fds[fdIdx].revents) + " on socket: " +
+            std::to_string(fds[fdIdx].fd));
+      }
+      // Now query the socket that has the event
+      try {
+        query(fds[fdIdx].fd);
+      } catch (...) {
+        // There was an error when processing query. Probably an exception
+        // occurred in recv/send what would indicate that socket on the other
+        // side has been closed. If the closing was due to normal exit, then
+        // the store should continue executing. Otherwise, if it was different
+        // exception, other connections will get an exception once they try to
+        // use the store. We will go ahead and close this connection whenever
+        // we hit an exception here.
+        ::close(fds[fdIdx].fd);
+        fds.erase(fds.begin() + fdIdx);
+        sockets_.erase(sockets_.begin() + fdIdx - 2);
```

Comment:
You should remove the socket from the `waiting*` maps as well. Otherwise you'll have the deamon reply to already closed descriptors, and likely crash the whole process.

Reply:
Ok maybe it wouldn't crash the whole process, but it would catch the exception and disconnect the setting process, which is really confusing.